vscode-trace-extension/src/extension.ts

```diff
@@ -102,3 +124,38 @@ export function activate(context: vscode.ExtensionContext): void {
 export function deactivate(): void {
     traceLogger.disposeChannel();
 }
+
+async function startTraceServerIfAvailable(): Promise<void>{
+    const traceServerExtension = vscode.extensions.getExtension('tracecompass-community.vscode-trace-server');
+    if (!traceServerExtension || await isUp()) {
+        pendingOpen = undefined;
+        return;
+    }
+    const extensionId = 'vscode-trace-server';
+    vscode.commands.executeCommand(extensionId + '.start-if-stopped');
+}
+
+async function isUp() {
+    if (tspClientProvider) {
+        const health = await tspClientProvider.getTspClient().checkHealth();
+        const status = health.getModel()?.status;
+        return health.isOk() && status === 'UP';
+    }
+    return true;
+}
+
+async function openTrace(analysisProvider: AnalysisProvider, content: PendingOpen) {
+    pendingOpen = content;
+    const fileOpenHandler = fileHandler(analysisProvider);
+    const startPromise = startTraceServerIfAvailable();
+    startPromise.then(() => {
+        if (pendingOpen === undefined) {
+            fileOpenHandler(content.context, content.file);
```

Comment:
The code handling `pendingOpen` is a bit odd. The only time there's a positive reference is in the handler for `serverStatus.started`, where something whose opening is pending will be opened. But the only place that `pendingOpen` is set is here in the `openTrace` function, where we insist on opening whatever's passed in, but only(?) if someone has set `pendingOpen` to `undefined` in the meantime. It basically seems to be aimed at two scenarios: if the server actually requires starting, then the server-starting plugin will invoke `serverStatus.started` when it starts and then the trace will be opened in that handler; alternatively, if the server is already up, then `pendingOpen` gets set to `undefined`, and `openTrace` handles the actual opening, but a piece of global state seems like a fragile way to handle those options.

Some possible solutions:
 - Ideally, the handler for `vscode-trace-server.start-if-stopped` wouldn't return until it had started the server or knew it couldn't. Then the code would just look like:
 ```
await startTraceServerIfAvailable();
if (await isUp()) {
    fileOpenHandler(content.context, content.file);
}
```
- Rather than adding `pendingOpen` to the global context, you could replace the command handler for `serverStatus.started` with one that performed the opening. That would have the advantage of at least keeping the variable local.

Reply:
Thank you very much for the review!

> The code handling `pendingOpen` is a bit odd. The only time there's a positive reference is in the handler for `serverStatus.started`, where something whose opening is pending will be opened. But the only place that `pendingOpen` is set is here in the `openTrace` function, where we insist on opening whatever's passed in, but only(?) if someone has set `pendingOpen` to `undefined` in the meantime. It basically seems to be aimed at two scenarios: if the server actually requires starting, then the server-starting plugin will invoke `serverStatus.started` when it starts and then the trace will be opened in that handler; alternatively, if the server is already up, then `pendingOpen` gets set to `undefined`, and `openTrace` handles the actual opening, but a piece of global state seems like a fragile way to handle those options.
> 
> Some possible solutions:
> 
>     * Ideally, the handler for `vscode-trace-server.start-if-stopped` wouldn't return until it had started the server or knew it couldn't. Then the code would just look like:
> 
> 
> ```
> await startTraceServerIfAvailable();
> if (await isUp()) {
>    fileOpenHandler(content.context, content.file);
> }
> ```
That was my initial idea as well. However, I was not able to make it work that `startTraceServerIfAvailable` waits until the server is actually running. The corresponding call to `vscode.executeCommand(...)` always returned right away and the start procedure was still ongoing while fileOpenHandler() was called. This let to issues when opening a trace from the file explorer (without the open dialog). Then I thought about using the `serverStatus.started` notification as trigger for calling `fileOpenHandler()`. It works, but the implementation was not clean and I was not really happy about it either. 

I just revisited the implementation of the `vscode-trace-server.start-if-stopped` command handler and I noticed that the command handler was not async nor awaiting properly, which caused the issues I had. I'll upload a patch in the `vscode-trace-server` repository to fix this. After that I'll update this PR. The solution will be simpler and cleaner afterwards.

> 
>     * Rather than adding `pendingOpen` to the global context, you could replace the command handler for `serverStatus.started` with one that performed the opening. That would have the advantage of at least keeping the variable local.

With my planed updates this won't be necessary anymore.